src/Sulu/Component/Webspace/Manager/WebspaceManager.php

```diff
@@ -247,48 +252,89 @@ public function findUrlByResourceLocator(
 
-        $sameDomainUrls = [];
-        $fullMatchedUrls = [];
-        $partialMatchedUrls = [];
+        if (isset($this->portalUrlCache[$webspaceKey][$domain][$environment][$languageCode])) {
+            $portalUrl = $this->portalUrlCache[$webspaceKey][$domain][$environment][$languageCode];
+
+            if (!$portalUrl) {
+                return null;
+            }
+
+            return $this->createResourceLocatorUrl($portalUrl, $resourceLocator, $scheme);
+        }
+
+        $sameDomainUrl = null;
+        $fullMatchedUrl = null;
+        $partialMatchedUrl = null;
 
         $portals = $this->getWebspaceCollection()->getPortalInformations(
-            $environment,
-            [
+            $environment
+        );
+
+        foreach ($portals as $portalInformation) {
+            if (!\in_array($portalInformation->getType(), [
                 RequestAnalyzerInterface::MATCH_TYPE_FULL,
                 RequestAnalyzerInterface::MATCH_TYPE_PARTIAL,
                 RequestAnalyzerInterface::MATCH_TYPE_REDIRECT,
-            ]
-        );
+            ])) {
+                continue;
+            }
+
+            $sameWebspace = null === $webspaceKey || $portalInformation->getWebspace()->getKey() === $webspaceKey;
+
+            if (!$sameWebspace) {
+                continue;
+            }
+
+            $portalLocalization = $portalInformation->getLocalization();
 
-        foreach ($portals as $portalInformation) {
             $sameLocalization = (
-                null === $portalInformation->getLocalization()
-                || $portalInformation->getLocalization()->getLocale() === $languageCode
+                null === $portalLocalization
+                || $portalLocalization->getLocale() === $languageCode
             );
-            $sameWebspace = null === $webspaceKey || $portalInformation->getWebspace()->getKey() === $webspaceKey;
-            $url = $this->createResourceLocatorUrl($portalInformation->getUrl(), $resourceLocator, $scheme);
-            if ($sameLocalization && $sameWebspace) {
-                if (RequestAnalyzerInterface::MATCH_TYPE_FULL === $portalInformation->getType()) {
-                    if ($this->isFromDomain($url, $domain)) {
-                        if ($portalInformation->isMain()) {
-                            \array_unshift($sameDomainUrls, $url);
-                        } else {
-                            $sameDomainUrls[] = $url;
-                        }
-                    } elseif ($portalInformation->isMain()) {
-                        \array_unshift($fullMatchedUrls, $url);
-                    } else {
-                        $fullMatchedUrls[] = $url;
+
+            if (!$sameLocalization) {
+                continue;
+            }
+
+            $portalUrl = $portalInformation->getUrl();
+            if (RequestAnalyzerInterface::MATCH_TYPE_FULL === $portalInformation->getType()) {
+                if ($this->isFromDomain('http://' . $portalUrl, $domain)) {
```

Comment:
not sure if we need to do this for `https` too?

Reply:
the scheme is not important for this call it just need to check if its the `sameDomain` or not.